src/realm/index_string.cpp

```diff
@@ -357,46 +358,44 @@ void IndexArray::index_string_all_ins(StringData value, IntegerColumn& result, C
     struct WorkItem {
         const char* header;
         size_t string_offset;
-        int permutation; // -1 == unexpanded, otherwise 0-15
+        key_type key;
     };
     using WorkList = std::vector<WorkItem>;
     WorkList work_list;
 
-    const char* top_header = get_header_from_data(m_data);
-    work_list.push_back({top_header, 0, -1});
-
     const util::Optional<std::string> upper_value = case_map(value, true);
     const util::Optional<std::string> lower_value = case_map(value, false);
 
+    auto generate_search_keys = [&work_list, &upper_value, &lower_value](const char* header,
+                                                                         size_t string_offset) -> void {
+        std::set<key_type> keys_seen;
+        constexpr int num_permutations = 1 << sizeof(key_type); // 4 bytes gives up to 16 search keys
+        const key_type upper_key = StringIndex::create_key(upper_value, string_offset);
+        const key_type lower_key = StringIndex::create_key(lower_value, string_offset);
+        for (int p = 0; p < num_permutations; ++p) {
+            // this might still be incorrect due to unicode characters crossing the 4 byte key size
+            const key_type key = generate_key(upper_key, lower_key, p);
+            const bool new_key = keys_seen.insert(key).second;
```

Comment:
When we spoke yesterday it sounded like you were leaning towards just disabling the use of the index for case-insensitive comparisons until we revisit our approach to FTS. That would take us back to the normal, well-tested code path.

Are we sufficiently confident in our testing of this fix that we think it's preferable to forge ahead with using the index rather than disabling it? Your earlier comment (https://github.com/realm/realm-core/pull/2578#issuecomment-297989879) implies we're not at that point. If we're not at that point, how do we intend to increase our confidence and is it worth spending the time on  that when we intend to take a different approach to this problem in the future?